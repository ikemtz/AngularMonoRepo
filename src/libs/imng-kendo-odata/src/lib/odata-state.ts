import { State, FilterDescriptor, CompositeFilterDescriptor, SortDescriptor } from '@progress/kendo-data-query';
import { IdType } from 'imng-nrsrx-client-utils';

export interface ODataState extends State {
  expanders?: Array<Expander | string>;
  selectors?: string[];
  inFilters?: InFilter[];
  childFilters?: CompositeChildFilterDescriptor;
  count?: boolean;
  transformations?: string;
  compute?: Array<Computation | string>;
}
export interface Computation {
  /** This can also be a static value */
  fieldA: string | number;
  /** This can also be a static value */
  fieldB: string | number;
  operator: 'mul' | 'div' | 'add' | 'sub' | 'mod' | string;
  /** This MUST differ from the names of declared or dynamic properties of the identified resources. */
  alias: string;
}

export interface InFilter {
  /** This value will default to 'and' in cases where there are additional filters specified. */
  logic?: 'or' | 'and';
  field: string;
  values: IdType[];
}

export interface CompositeChildFilterDescriptor extends CompositeFilterDescriptor {
  filters: Array<CompositeChildFilterDescriptor | ChildFilterDescriptor>;
}

export interface ChildFilterDescriptor extends FilterDescriptor {
  /**  This value will default to 'and' in cases where there are additional filters specified. */
  childTableNavigationProperty: string;
  linqOperation: 'all' | 'any';
}

export interface Expander {
  table: string;
  selectors?: string[];
  expander?: string | Expander;
  filter?: CompositeFilterDescriptor;
  sort?: Array<SortDescriptor>;
}
export function isExpander(source: string | Expander): source is Expander {
  return !!(source as Expander)?.table;
}
export function isComputation(source: string | Computation): source is Computation {
  return !!(source as Computation)?.operator;
}
