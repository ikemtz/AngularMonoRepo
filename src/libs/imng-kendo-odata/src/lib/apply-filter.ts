import { FilterDescriptor } from '@progress/kendo-data-query';
import { findMatchingFilters } from './find-filter';
import { ODataState } from './odata-state';

export function applyFilter(odataState: ODataState, filter: FilterDescriptor): ODataState {
  if (!odataState.filter) {
    odataState.filter = { logic: 'and', filters: [] };
  }
  const matchedFilter = findMatchingFilters(odataState, filter.field as string);
  if (!matchedFilter) {
    odataState.filter.filters.push({
      logic: 'and',
      filters: [filter]
    });
  }
  else {
    matchedFilter.operator = 'eq';
    matchedFilter.value = filter.value;
  }
  return odataState;
}
