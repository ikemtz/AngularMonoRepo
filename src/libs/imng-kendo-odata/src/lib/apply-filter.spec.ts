import { FilterDescriptor } from '@progress/kendo-data-query';
import { applyFilter } from './apply-filter';
import { ODataState } from './odata-state';

describe('applyFilter', () => {
  const genFilter = (): FilterDescriptor => ({ field: 'ABC', operator: 'eq', value: 2 });
  it('should should initialize', () => {
    const result = applyFilter({}, genFilter());
    expect(result).toMatchSnapshot();
  });

  it('should should override values', () => {
    const state: ODataState = {
      filter: {
        logic: 'and',
        filters: [
          {
            logic: 'and',
            filters: [{ field: 'ABC', operator: 'lte', value: 6 }]
          }
        ]
      }
    };

    const result = applyFilter(state, genFilter());
    expect(result).toMatchSnapshot();
  });
});
