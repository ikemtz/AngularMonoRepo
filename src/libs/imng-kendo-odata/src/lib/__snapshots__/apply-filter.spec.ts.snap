// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyFilter should should initialize 1`] = `
Object {
  "filter": Object {
    "filters": Array [
      Object {
        "filters": Array [
          Object {
            "field": "ABC",
            "operator": "eq",
            "value": 2,
          },
        ],
        "logic": "and",
      },
    ],
    "logic": "and",
  },
}
`;

exports[`applyFilter should should override values 1`] = `
Object {
  "filter": Object {
    "filters": Array [
      Object {
        "filters": Array [
          Object {
            "field": "ABC",
            "operator": "eq",
            "value": 2,
          },
        ],
        "logic": "and",
      },
    ],
    "logic": "and",
  },
}
`;
